pages/book/security-best-practices.mdx

```diff
@@ -0,0 +1,144 @@
+# Security Best Practices
+
+In Tact smart contracts there are various anti-patterns and potential attacks developers need to be cautious of. These can impact the security, efficiency, and correctness of the contracts. Below, are covered both common anti-patterns and attacks specific to Tact smart contracts.
+
+## Send private data on chain
+
+Everything is stored on the blockchain. Remember that even if no private data is saved in the transaction, anyone can still get this data.
+
+## Incorrect use of signed/unsigned integer
+
+The following is an example of incorrect use of a signed integer. In the `Vote` message, the `votes` type is `int32`. This can lead to fraud when the attacker sends the negative number of votes instead of positive one.
+
+Invalid example:
+
+```tact
+message Vote {
+    votes: Int as int32;
+}
+
+contract Sample {
+    votes: Int as uint32 = 0;
+
+    receive(msg: Vote) {
+        self.votes += msg.votes;
+    }
+}
+```
+
+Unsigned integers are safer because they will error out if an overflow occurs, and signed integers are only used when they are really needed.
+
+
+## Invalid throw values
+
+Each time the TVM execution stops normally, it stops with exit codes `0` or `1`.
+Although it is done automatically, TVM execution can be interrupted directly in an unexpected way if exit codes `0` and `1` are thrown directly by either `throw(0)` or `throw(1)` command.
+
+Invalid example:
+
+```tact
+throw(0);
+throw(1);
+```
+
+## Insecure random numbers
+
+Note that [`random()`](/ref/core-random#random) function is pseudo-random and depends on [logical time](https://docs.ton.org/develop/smart-contracts/guidelines/message-delivery-guarantees#what-is-a-logical-time). A hacker can predict random by brute-forcing the logical time in the current block.
+
+Invalid example:
+
+```tact
+if (random(1, 10) == 7) {
+  ... send reward ...
+}
+```
+
+## Optimized message handling
+
+The frontend should handle string parsing from human-friendly formats into machine-readable binary structures.
+This ensures that only optimized, compact messages are sent to the blockchain, reducing computational and storage costs.
+By doing this, you avoid overloading the blockchain with unnecessary processing, thereby improving efficiency.
```

Comment:
Blockchain will be fine, as well as the smart contract, haha :)
But fees (both gas and forward) will be higher when using some human readable strings instead of binary structs.